### PR TITLE
Issue #5779: fixing no violation on parenthesis in if statement betwe…

### LINKED ISSUE #5779: UnnecessaryParentheses: no violation on parenthesis in if statement between || and &&
Identified at https://github.com/checkstyle/checkstyle/pull/5778#discussion_r186302647 ,

````
$ cat TestClass.java
public class TestClass {
    void method(String sectionName) {
        if ("Content".equals(sectionName) || "Overview".equals(sectionName)
                // suppression list
                || (!"AbbreviationAsWordInName".equals(sectionName)
                        && !"AbstractClassName".equals(sectionName)
                        )) {
             continue;
         }
    }
}

$ cat TestConfig.xml
<?xml version="1.0"?>
<!DOCTYPE module PUBLIC
          "-//Puppy Crawl//DTD Check Configuration 1.3//EN"
          "http://www.puppycrawl.com/dtds/configuration_1_3.dtd">

<module name="Checker">
    <property name="charset" value="UTF-8"/>

    <module name="TreeWalker">
<module name="UnnecessaryParentheses" />
    </module>
</module>

$ java -jar checkstyle-8.10-all.jar -c TestConfig.xml TestClass.java
Starting audit...
Audit done.
````

Every other CI complains about parenthesis at line 5 except our check. There should be a violation there.

### PR DESCRIPTION
**Pull request by Vyom-Yadav**
Resolves #5779: fixing no violation on parenthesis in if statement between || and &&
Diff Regression projects: https://gist.githubusercontent.com/Vyom-Yadav/91807e6244cff60698d9e33e32ba2d51/raw/85355576c05c80e8b08752a62ba3a4076909d266/projects-to-test-on.properties
Diff Regression config: https://gist.githubusercontent.com/Vyom-Yadav/98dceb63a79f4833e85fff9b2e1464a6/raw/7a57a856b6657d8606276270a6ca3276d102ef65/my_checks.xml

### CODE REVIEW

**Comment by Vyom-Yadav** [comment]
@strkkk @rnveach I constantly face this issue
```
All check input files in the 'src/(it|test)/resources*/' folder must                    have trailing comments "// ok" and/or "// violation". [StyleValidationCommentInInputs]
```
I have added `// ok` or `// violation` comment to almost every single line

**Comment by strkkk** [comment]
@Vyom-Yadav there is no need to add this to every line. 
You only need to add violation comment to lines with violations. In case there is no violation, add ok comment.

**Comment by Vyom-Yadav** [comment]
> @Vyom-Yadav there is no need to add this to every line.
> You only need to add violation comment to lines with violations. In case there is no violation, add ok comment.
> 
> Purpose of this check is to report violations on files without any violation/ok comments at all

It is still showing the same error, I added violation comment in front of every line that has a violation and ok in front of lines which don't. I tried many times in my local but couldn't find a solution for it. Can you please see the resource file and point out where I am going wrong.

**Comment by strkkk** [comment]
I remember that someone else had similar problem on their local machine, but it disappeared when they push changes to github

**Comment by Vyom-Yadav** [comment]
Github, generate report

**Comment by Vyom-Yadav** [comment]
Github, generate report

**Comment by Vyom-Yadav** [comment]
> https://checkstyle-diff-reports.s3.us-east-2.amazonaws.com/noViolationUnnecessaryParentheses_2021150215/reports/diff/index.html

@strkk I think there is a problem, will workaround.

**Comment by Vyom-Yadav** [comment]
@strkkk Many violation I am seeing are just using extra parenthesis for clarity purpose and even if we remove them, the logic wont be affected. Please review it once.

**Comment by strkkk** [comment]
@Vyom-Yadav wercker is failing since there violations in Orekit (see below).
Please send PR for them to fix it.

Please also add test cases with booleans like https://checkstyle-diff-reports.s3.us-east-2.amazonaws.com/noViolationUnnecessaryParentheses_2021150215/reports/diff/Orekit/xref/home/runner/work/checkstyle/checkstyle/contribution/checkstyle-tester/repositories/Orekit/src/main/java/org/orekit/attitudes/AttitudesSequence.java.html#L347

and more complicated cases like https://checkstyle-diff-reports.s3.us-east-2.amazonaws.com/noViolationUnnecessaryParentheses_2021150215/reports/diff/lombok-ast/xref/home/runner/work/checkstyle/checkstyle/contribution/checkstyle-tester/repositories/lombok-ast/src/main/lombok/ast/grammar/Source.java.html#L481

```
[INFO] There are 53 errors reported by Checkstyle 8.41-SNAPSHOT with /pipeline/source/.ci-temp/Orekit/checkstyle.xml ruleset.
[ERROR] src/main/java/org/orekit/propagation/semianalytical/dsst/forces/FieldDSSTThirdBodyContext.java:[192,53] (coding) UnnecessaryParentheses: Unnecessary parentheses around expression.
[ERROR] src/main/java/org/orekit/propagation/semianalytical/dsst/forces/DSSTThirdBodyContext.java:[183,43] (coding) UnnecessaryParentheses: Unnecessary parentheses around expression.
[ERROR] src/main/java/org/orekit/propagation/semianalytical/dsst/utilities/NewcombOperators.java:[473,37] (coding) UnnecessaryParentheses: Unnecessary parentheses around expression.
[ERROR] src/main/java/org/orekit/propagation/semianalytical/dsst/utilities/NewcombOperators.java:[553,34] (coding) UnnecessaryParentheses: Unnecessary parentheses around expression.
[ERROR] src/main/java/org/orekit/propagation/semianalytical/dsst/utilities/CoefficientsFactory.java:[323,34] (coding) UnnecessaryParentheses: Unnecessary parentheses around expression.
[ERROR] src/main/java/org/orekit/propagation/semianalytical/dsst/DSSTPartialDerivativesEquations.java:[214,13] (coding) UnnecessaryParentheses: Unnecessary parentheses around expression.
[ERROR] src/main/java/org/orekit/propagation/semianalytical/dsst/DSSTPartialDerivativesEquations.java:[215,13] (coding) UnnecessaryParentheses: Unnecessary parentheses around expression.
[ERROR] src/main/java/org/orekit/propagation/numerical/PartialDerivativesEquations.java:[217,13] (coding) UnnecessaryParentheses: Unnecessary parentheses around expression.
[ERROR] src/main/java/org/orekit/propagation/numerical/PartialDerivativesEquations.java:[218,13] (coding) UnnecessaryParentheses: Unnecessary parentheses around expression.
[ERROR] src/main/java/org/orekit/propagation/events/EventState.java:[253,41] (coding) UnnecessaryParentheses: Unnecessary parentheses around expression.
[ERROR] src/main/java/org/orekit/propagation/events/EventState.java:[253,67] (coding) UnnecessaryParentheses: Unnecessary parentheses around expression.
[ERROR] src/main/java/org/orekit/propagation/events/EventState.java:[341,19] (coding) UnnecessaryParentheses: Unnecessary parentheses around expression.
[ERROR] src/main/java/org/orekit/propagation/events/EventState.java:[342,19] (coding) UnnecessaryParentheses: Unnecessary parentheses around expression.
[ERROR] src/main/java/org/orekit/propagation/events/FieldEventState.java:[260,61] (coding) UnnecessaryParentheses: Unnecessary parentheses around expression.
[ERROR] src/main/java/org/orekit/propagation/events/FieldEventState.java:[260,107] (coding) UnnecessaryParentheses: Unnecessary parentheses around expression.
[ERROR] src/main/java/org/orekit/propagation/events/FieldEventState.java:[346,19] (coding) UnnecessaryParentheses: Unnecessary parentheses around expression.
[ERROR] src/main/java/org/orekit/propagation/events/FieldEventState.java:[347,19] (coding) UnnecessaryParentheses: Unnecessary parentheses around expression.
[ERROR] src/main/java/org/orekit/estimation/measurements/gnss/IntegerLeastSquareSolution.java:[65,32] (coding) UnnecessaryParentheses: Unnecessary parentheses around expression.
[ERROR] src/main/java/org/orekit/estimation/leastsquares/MeasurementHandler.java:[76,17] (coding) UnnecessaryParentheses: Unnecessary parentheses around expression.
[ERROR] src/main/java/org/orekit/estimation/leastsquares/MeasurementHandler.java:[77,17] (coding) UnnecessaryParentheses: Unnecessary parentheses around expression.
[ERROR] src/main/java/org/orekit/orbits/KeplerianOrbit.java:[723,17] (coding) UnnecessaryParentheses: Unnecessary parentheses around expression.
[ERROR] src/main/java/org/orekit/orbits/CartesianOrbit.java:[623,17] (coding) UnnecessaryParentheses: Unnecessary parentheses around expression.
[ERROR] src/main/java/org/orekit/orbits/FieldKeplerianOrbit.java:[803,17] (coding) UnnecessaryParentheses: Unnecessary parentheses around expression.
[ERROR] src/main/java/org/orekit/orbits/FieldCartesianOrbit.java:[650,17] (coding) UnnecessaryParentheses: Unnecessary parentheses around expression.
[ERROR] src/main/java/org/orekit/attitudes/AttitudesSequence.java:[410,18] (coding) UnnecessaryParentheses: Unnecessary parentheses around expression.
[ERROR] src/main/java/org/orekit/attitudes/AttitudesSequence.java:[410,54] (coding) UnnecessaryParentheses: Unnecessary parentheses around expression.
[ERROR] src/main/java/org/orekit/gnss/RinexLoader.java:[486,45] (coding) UnnecessaryParentheses: Unnecessary parentheses around expression.
[ERROR] src/main/java/org/orekit/gnss/RinexLoader.java:[487,45] (coding) UnnecessaryParentheses: Unnecessary parentheses around expression.
[ERROR] src/main/java/org/orekit/gnss/RinexLoader.java:[950,45] (coding) UnnecessaryParentheses: Unnecessary parentheses around expression.
[ERROR] src/main/java/org/orekit/gnss/RinexLoader.java:[951,45] (coding) UnnecessaryParentheses: Unnecessary parentheses around expression.
[ERROR] src/main/java/org/orekit/models/earth/tessellation/Mesh.java:[428,21] (coding) UnnecessaryParentheses: Unnecessary parentheses around expression.
[ERROR] src/main/java/org/orekit/models/earth/tessellation/Mesh.java:[428,47] (coding) UnnecessaryParentheses: Unnecessary parentheses around expression.
[ERROR] src/main/java/org/orekit/time/DateComponents.java:[394,57] (coding) UnnecessaryParentheses: Unnecessary parentheses around expression.
[ERROR] src/main/java/org/orekit/time/FieldAbsoluteDate.java:[1405,31] (coding) UnnecessaryParentheses: Unnecessary parentheses around expression.
[ERROR] src/main/java/org/orekit/time/AbsoluteDate.java:[1224,31] (coding) UnnecessaryParentheses: Unnecessary parentheses around expression.
[ERROR] src/main/java/org/orekit/forces/drag/AbstractDragForceModel.java:[74,27] (coding) UnnecessaryParentheses: Unnecessary parentheses around expression.
[ERROR] src/main/java/org/orekit/forces/gravity/potential/ICGEMFormatReader.java:[286,25] (coding) UnnecessaryParentheses: Unnecessary parentheses around expression.
[ERROR] src/main/java/org/orekit/forces/gravity/potential/ICGEMFormatReader.java:[286,68] (coding) UnnecessaryParentheses: Unnecessary parentheses around expression.
[ERROR] src/main/java/org/orekit/frames/RapidDataAndPredictionXMLLoader.java:[134,48] (coding) UnnecessaryParentheses: Unnecessary parentheses around expression.
[ERROR] src/main/java/org/orekit/bodies/JPLEphemeridesLoader.java:[543,17] (coding) UnnecessaryParentheses: Unnecessary parentheses around expression.
[ERROR] src/main/java/org/orekit/bodies/JPLEphemeridesLoader.java:[544,21] (coding) UnnecessaryParentheses: Unnecessary parentheses around expression.
[ERROR] src/main/java/org/orekit/bodies/JPLEphemeridesLoader.java:[545,21] (coding) UnnecessaryParentheses: Unnecessary parentheses around expression.
[ERROR] src/main/java/org/orekit/bodies/JPLEphemeridesLoader.java:[546,21] (coding) UnnecessaryParentheses: Unnecessary parentheses around expression.
[ERROR] src/main/java/org/orekit/bodies/JPLEphemeridesLoader.java:[547,21] (coding) UnnecessaryParentheses: Unnecessary parentheses around expression.
[ERROR] src/main/java/org/orekit/bodies/JPLEphemeridesLoader.java:[548,21] (coding) UnnecessaryParentheses: Unnecessary parentheses around expression.
[ERROR] src/main/java/org/orekit/bodies/JPLEphemeridesLoader.java:[549,21] (coding) UnnecessaryParentheses: Unnecessary parentheses around expression.
[ERROR] src/main/java/org/orekit/bodies/JPLEphemeridesLoader.java:[550,21] (coding) UnnecessaryParentheses: Unnecessary parentheses around expression.
[ERROR] src/main/java/org/orekit/bodies/JPLEphemeridesLoader.java:[551,21] (coding) UnnecessaryParentheses: Unnecessary parentheses around expression.
[ERROR] src/main/java/org/orekit/bodies/JPLEphemeridesLoader.java:[552,21] (coding) UnnecessaryParentheses: Unnecessary parentheses around expression.
[ERROR] src/main/java/org/orekit/bodies/JPLEphemeridesLoader.java:[553,21] (coding) UnnecessaryParentheses: Unnecessary parentheses around expression.
[ERROR] src/main/java/org/orekit/files/ccsds/ODMMetaData.java:[273,13] (coding) UnnecessaryParentheses: Unnecessary parentheses around expression.
[ERROR] src/main/java/org/orekit/files/ccsds/ODMMetaData.java:[274,17] (coding) UnnecessaryParentheses: Unnecessary parentheses around expression.
[ERROR] src/main/java/org/orekit/files/ccsds/ODMMetaData.java:[275,17] (coding) UnnecessaryParentheses: Unnecessary parentheses around expression.
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
```

File: src/test/resources/com/puppycrawl/tools/checkstyle/checks/coding/unnecessaryparentheses/InputUnnecessaryParenthesesIfStatement.java [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+        if ((x == y) || (x != z)) { // ok
+            return;
+        }
+
+        if (((x == y) || (x != z))) { // violation, unnecessary parenthesis
+            return;
+        }
+
+        if ((Integer.valueOf(x) instanceof Integer) // violation, unnecessary parenthesis
+                || Integer.valueOf(y) instanceof Integer) { // ok
```
Comment (rnveach) [inline src/test/resources/com/puppycrawl/tools/checkstyle/checks/coding/unnecessaryparentheses/InputUnnecessaryParenthesesIfStatement.java:46]: What is the requirements to pick something up for a violation?

This one has a violation which is following the pattern `(a) || b`, but

File: src/test/resources/com/puppycrawl/tools/checkstyle/checks/coding/unnecessaryparentheses/InputUnnecessaryParenthesesIfStatement.java [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+
+        if ((x < y) && (x > z)) { // ok
+            return;
+        }
+
+        if (((x < y) && (x > z))) { // violation, unnecessary parenthesis
+            return;
+        }
+
+        if (!(x <= y) || (x >= z)) { // ok
```
Comment (rnveach) [inline src/test/resources/com/puppycrawl/tools/checkstyle/checks/coding/unnecessaryparentheses/InputUnnecessaryParenthesesIfStatement.java:33]: this is following the pattern `!a || (b)` and does not have a violation.

I am confused by these and would like some clarification.

**Comment by Vyom-Yadav** [comment]
@rnveach After thinking for some time and evaluating the use of parenthesis, I have a suggestion. Also, there are some discrepancies in logic which I will fix.
```
public class TestClass {
    void method(String sectionName) {
        if ("Content".equals(sectionName) || "Overview".equals(sectionName)
                // suppression list
                || (!"AbbreviationAsWordInName".equals(sectionName)
                        && !"AbstractClassName".equals(sectionName)
                        )) {
             continue;
         }
    }
}
```
In the issue, you mentioned that there was a missing violation on the 5th line (Unnecessary parenthesis). 
Let's breakdown this if statement. It basically consists of 3 parts-

1st. `if ("Content".equals(sectionName))`
2nd.` if ("Overview".equals(sectionName))`
3rd. `if (!"AbbreviationAsWordInName".equals(sectionName) && "AbstractClassName".equals(sectionName) )`

While it may be correct to leave the parenthesis out on the 3rd condition as after the `|| `, the code will be evaluated as a single condition - for anyone looking at the 3rd condition these 'redundant' parenthesis actually provide a little clarity as to what forms part of this condition. There is no doubt that both of these `"AbbreviationAsWordInName".equals(sectionName)` and `"AbstractClassName".equals(sectionName)` have to be meet in order to evaluate to true.

If you left this parenthesis out, one could get confused about whether these two form part of a single condition or not.
I guess at the end of the day (at least in my opinion) it's better to clearly segregate logical conditions so they are easier to understand but also extra parentheses, just like indentations, whitespace, and other styles standards, help visually organize the code in a logical way.
I will proceed as you say.

**Comment by rnveach** [comment]
> I guess at the end of the day (at least in my opinion) it's better to clearly segregate logical conditions so they are easier to understand but also extra parentheses, just like indentations, whitespace, and other styles standards, help visually organize the code in a logical way.

> I will proceed as you say.

If you are unsure about what would constitute a violation or not, then I would ask @romani to clarify what he wants to be done here.

**Comment by rnveach** [comment]
> I guess at the end of the day (at least in my opinion) it's better to clearly segregate logical conditions so they are easier to understand but also extra parentheses, just like indentations, whitespace, and other styles standards, help visually organize the code in a logical way.

I agree. I write code with `()`s around all my conditions in if statement outside of Checkstyle to clarify reading order.

The Checkstyle project, however, has other CI rules in place to remove what it probably calls "redundant" `()`s. These are defined as `()`s that don't change the flow of the default behavior for the order of operations followed for `&&`s and `||`s. So if a `()` does change the flow of the default order, then it isn't unnecessary. This issue was created from one of those redundancies that was found in another PR.

> I will proceed as you say.

If you are unsure about what would constitute a violation or not, and what scope this issue/PR should be focusing on, then you should be asking for this clarification. I would ask @romani to clarify what he feels should be done here so it is clear.

**Comment by Vyom-Yadav** [comment]
@romani, can you point out ( with examples ) where we need parenthesis and where we don't. 
```
if (!(x <= y) || (x >= z)) { // ok
```
We want whitespace around `x<= y` ( as we are negating it ) and not around `x>=z` right?

**Comment by Vyom-Yadav** [comment]
Github, generate report

**Comment by Vyom-Yadav** [comment]
@rnveach I saw the report, and it's working as supposed. One thing
```
int x = 5;
int y = 6;
int z;
z = (x>y) ? x:y; // no violation
```
I will send another PR for this as this issue was related to `&&`,`||` inside `if` statements. 
Another thing-
https://checkstyle-diff-reports.s3.us-east-2.amazonaws.com/noViolationUnnecessaryParentheses_2021121301/reports/diff/lombok-ast/xref/home/runner/work/checkstyle/checkstyle/contribution/checkstyle-tester/repositories/lombok-ast/test/resources/performance/extra166y/PAS.java.html#L1491
There is no violation on line 1492 as expected.

**Comment by rnveach** [comment]
> NoErrorTest - Orekit
[INFO] --- maven-checkstyle-plugin:3.1.1:check (default-cli) @ orekit ---
[INFO] There are 304 errors reported by Checkstyle 8.41-SNAPSHOT with /pipeline/source/.ci-temp/Orekit/checkstyle.xml ruleset.
[ERROR] src/main/java/org/orekit/frames/UpdatableFrame.java:[149,35] (coding) UnnecessaryParentheses: Unnecessary parentheses around expression.

**Comment by Vyom-Yadav** [comment]
> > NoErrorTest - Orekit
> > [INFO] --- maven-checkstyle-plugin:3.1.1:check (default-cli) @ orekit ---
> > [INFO] There are 304 errors reported by Checkstyle 8.41-SNAPSHOT with /pipeline/source/.ci-temp/Orekit/checkstyle.xml ruleset.
> > [ERROR] src/main/java/org/orekit/frames/UpdatableFrame.java:[149,35] (coding) UnnecessaryParentheses: Unnecessary parentheses around expression.

Just this one? I will send another PR to orekit for this one.

**Comment by rnveach** [comment]
Please read the statement above a bit closer.

There are **304** errors.

**Comment by Vyom-Yadav** [comment]
@rnveach Oops, missed that, I want to confirm before sending a PR thats these changes are ok or not.

File: src/test/resources/com/puppycrawl/tools/checkstyle/checks/coding/unnecessaryparentheses/InputUnnecessaryParenthesesIfStatement.java (hunk 1 above)
Comment (Vyom-Yadav) [reply src/test/resources/com/puppycrawl/tools/checkstyle/checks/coding/unnecessaryparentheses/InputUnnecessaryParenthesesIfStatement.java:46]: fixed, now a violation

File: src/test/resources/com/puppycrawl/tools/checkstyle/checks/coding/unnecessaryparentheses/InputUnnecessaryParenthesesIfStatement.java (hunk 2 above)
Comment (Vyom-Yadav) [reply src/test/resources/com/puppycrawl/tools/checkstyle/checks/coding/unnecessaryparentheses/InputUnnecessaryParenthesesIfStatement.java]: fixed now a violation

**Comment by Vyom-Yadav** [comment]
@rnveach @strkkk ping

**Comment by rnveach** [comment]
i don't understand what I am being pinged for.

Violations needs to be looked at if they are correct or not. If they are correct, see if the project will make the changes shown by the regression by contributing a PR to their project. If they don't then we can discuss more.

**Comment by Vyom-Yadav** [comment]
> i don't understand what I am being pinged for.
> 
> Violations needs to be looked at if they are correct or not. If they are correct, see if the project will make the changes shown by the regression by contributing a PR to their project. If they don't then we can discuss more.

I don't have access to the wrecker, I will need the whole wrecker log for making changes in orekit project.

**Comment by strkkk** [comment]
I shared output in gitter

**Comment by Vyom-Yadav** [comment]
@strkk While doing changes in orekit project, I noticed one issue-
```
int a = 5;
boolean b = true;
if ((!b) && a<7) { // no violation reported 
return;
}
```
I fixed it so now it reports a violation.
Note - It won't affect `!(a && b)` as without parenthesis the logic will change.

**Comment by Vyom-Yadav** [comment]
Github, generate report

**Comment by rnveach** [comment]
> I don't have access to the wrecker, I will need the whole wrecker log for making changes in orekit project.

Wercker commands can be run locally to identify issues, and confirm they are fixed as you are making changes. 
https://github.com/checkstyle/checkstyle/blob/db7215dc8b80fa99c27250c0548e9581f0f40208/wercker.yml#L80

This is why I provided project name and limited output.

**Comment by Vyom-Yadav** [comment]
Closed, will do in a new PR.

**Pull request closed without merging**